tests/unit/base/CrossValidation_unittest.cc

```diff
@@ -0,0 +1,204 @@
+/*
+ * This software is distributed under BSD 3-clause license (see LICENSE file).
+ *
+ * Authors: Heiko Strathmann
+ */
+
+#include <gtest/gtest.h>
+#include <shogun/evaluation/CrossValidation.h>
+#include <shogun/evaluation/CrossValidationSplitting.h>
+#include <shogun/evaluation/ContingencyTableEvaluation.h>
+#include <shogun/evaluation/MulticlassAccuracy.h>
+#include <shogun/evaluation/MeanSquaredError.h>
+
+#include <shogun/kernel/GaussianKernel.h>
+#include <shogun/distance/EuclideanDistance.h>
+
+#include <shogun/classifier/svm/LibSVM.h>
+#include <shogun/classifier/Perceptron.h>
+#include <shogun/classifier/svm/LibLinear.h>
+#include <shogun/multiclass/MulticlassLibLinear.h>
+#include <shogun/multiclass/KNN.h>
+#include <shogun/regression/LinearRidgeRegression.h>
+#include <shogun/regression/KernelRidgeRegression.h>
+
+#include <shogun/mathematics/NormalDistribution.h>
+#include <shogun/mathematics/linalg/LinalgNamespace.h>
+
+
+
+using namespace shogun;
+
+
+#include <gtest/gtest.h>
+
+using namespace shogun;
+using namespace std;
+
+template <class T>
+class CrossValidationTests : public ::testing::Test
+{
+protected:
+	void SetUp()
+	{
+	}
+
+	void init()
+	{
+		machine = new T();
+		SG_REF(machine)
+
+		this->generate_data(this->machine->get_machine_problem_type());
+
+		if (auto* casted = dynamic_cast<CKernelMachine*>(machine))
+			casted->set_kernel(new CGaussianKernel());
+		if (auto* casted = dynamic_cast<CDistanceMachine*>(machine))
+			casted->set_distance(new CEuclideanDistance());
+
+		auto ss = new CCrossValidationSplitting(labels, 5);
+		CEvaluation* ec = nullptr;
+		switch (machine->get_machine_problem_type())
+		{
+		case PT_BINARY:
+			ec = new CAccuracyMeasure();
+			break;
+		case PT_MULTICLASS:
+			ec = new CMulticlassAccuracy();
+			break;
+		case PT_REGRESSION:
+			ec = new CMeanSquaredError();
+			break;
+		default:
+			SG_SNOTIMPLEMENTED
+			break;
+		}
+		cv = new CCrossValidation(machine, features, labels, ss, ec);
+		cv->set_num_runs(3);
+		SG_REF(cv);
+	}
+
+	void clean()
+	{
+		SG_UNREF(features)
+		SG_UNREF(labels)
+		SG_UNREF(machine)
+		SG_UNREF(cv);
+	}
+
+	void TearDown()
+	{
+	}
+
+	auto test_single_thread()
+	{
+		init();
+		this->cv->put("seed", 1);
+		get_global_parallel()->set_num_threads(1);
+		auto result = cv->evaluate()->get<float64_t>("mean");
+		clean();
+		return result;
+	}
+
+	auto test_multi_thread()
+	{
+		init();
+		this->cv->put("seed", 1);
+		get_global_parallel()->set_num_threads(4);
+		auto result = cv->evaluate()->get<float64_t>("mean");
+		clean();
+		return result;
+	}
+
+	void generate_data(EProblemType pt)
```

Comment:
I was using the test environments at first but most algorithms get accuracy of 1.0 or similar in there, so it is not that meaningful to compare the xvalidation results with and without threads.... so now some custom problems. Once this is merged, I will try to write a test framework that tests all machines (or sub-sets of those) on particular environments. But for now I think this is good